### PR TITLE
React/redux task detail page

### PR DESCRIPTION
**Pull request by kwm4385**
Includes new modals, redux components, and general use components like file browser, confirmation dialog, and more.

Also introduces the option to key a redux action and reducer. 

@tpetr @wolfd @Calvinp

### CODE REVIEW

File: SingularityUI/app/actions/api/base.es6 [hunk 1]
```diff
@@ -2,7 +2,7 @@ import fetch from 'isomorphic-fetch';
 
 const JSON_HEADERS = {'Content-Type': 'application/json', 'Accept': 'application/json'};
 
-export function buildJsonApiAction(actionName, httpMethod, opts={}) {
+export function buildJsonApiAction(actionName, httpMethod, opts={}, keyField=undefined) {
```
Comment (tpetr) [inline SingularityUI/app/actions/api/base.es6]: Asking out of curiosity -- why the decision to explicitly default keyField to `undefined`?

File: SingularityUI/app/actions/api/base.es6 [hunk 2]
```diff
@@ -40,18 +40,29 @@ export function buildApiAction(actionName, opts={}) {
     optsFunc = () => opts;
   }
 
+  if (typeof opts === 'object') {
+    optsFunc = () => opts;
```
Comment (tpetr) [inline SingularityUI/app/actions/api/base.es6]: won't this already be handled by the if block above?

File: SingularityUI/app/actions/api/base.es6 [hunk 3]
```diff
@@ -66,16 +77,16 @@ export function buildApiAction(actionName, opts={}) {
     return { type: CLEAR };
   }
 
-  function started() {
-    return { type: STARTED };
+  function started(key=undefined) {
+    return { type: STARTED, key: key };
```
Comment (tpetr) [inline SingularityUI/app/actions/api/base.es6]: you can just do `{ type: STARTED, key };`

File: SingularityUI/app/actions/api/task.es6 [hunk 4]
```diff
@@ -1,46 +1,7 @@
...
-}
-
-export function fetchTasksError(error, taskId) {
-  return { type: FETCH_TASK_ERROR, error: error, taskId: taskId };
-}
-
-export function fetchTaskSuccess(data, taskId) {
-  return { type: FETCH_TASK_SUCCESS, data: data, taskId: taskId };
-}
+export const KillAction = buildApiAction('KILL_TASK', (taskId) => {return {url: `/tasks/task/${taskId}`, method: 'DELETE'}});
```
Comment (tpetr) [inline SingularityUI/app/actions/api/task.es6]: `(taskId) => ({ blah, blah })`

File: SingularityUI/app/components/taskDetail/ShellCommandLauncher.jsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  renderStatusList() {
+    return (
+      <ul className="status-list">
+        <li className={this.state.commandAcked ? 'complete text-success' : 'waiting'}>
+          {!this.state.commandAcked ? <div className="page-loader loader-small" /> : <Glyphicon iconClass='ok' />} Command acknowledged...
+        </li>
+        <li className={this.state.commandStarted ? 'complete text-success' : 'waiting'}>
+          {!this.state.commandStarted ? <div className="page-loader loader-small" /> : <Glyphicon iconClass='ok' />} Command started...
+        </li>
+        <li className={this.state.commandFileExists ? 'complete text-success' : 'waiting'}>
```
Comment (tpetr) [inline SingularityUI/app/components/taskDetail/ShellCommandLauncher.jsx]: consider using `classNames()` for these

File: SingularityUI/app/actions/api/base.es6 (hunk 1 above)
Comment (kwm4385) [inline SingularityUI/app/actions/api/base.es6]: There's no functional reason, I decided to have that to make it more obvious to the reader that the parameter is explicitly optional.

File: SingularityUI/app/components/taskDetail/ShellCommandLauncher.jsx [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+  }
+
+  isDone(state) {
+    return classNames({'complete text-success': state}, {'waiting': !state});
+  }
+
+  renderStatusList() {
+    return (
+      <ul className="status-list">
+        <li className={this.isDone(this.state.commandAcked)}>
```
Comment (andyhuang91) [inline SingularityUI/app/components/taskDetail/ShellCommandLauncher.jsx]: maybe create a helper for each individual li

File: SingularityUI/app/components/taskDetail/TaskDetail.jsx [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+          onConfirm={this.killTask.bind(this)}
+          buttonStyle="danger"
+        />
+      <a className="btn btn-danger" onClick={() => this.refs.confirmKillTask.show()}>
+          {removeText}
+        </a>
+      </span>
+    );
+    const terminationAlert = t.isStillRunning && !cleanup && t.isCleaning && (
+      <div className="alert alert-warning" role="alert">
```
Comment (andyhuang91) [inline SingularityUI/app/components/taskDetail/TaskDetail.jsx]: Could use the react-bootstrap Alert component here.

File: SingularityUI/app/components/taskDetail/TaskDetail.jsx [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+    resourceUsage: state.api.taskResourceUsage.data,
+    cpuTimestamp: state.api.taskResourceUsage.data.timestamp,
+    s3Logs: state.api.taskS3Logs.data,
+    deploy: state.api.deploy.data,
+    pendingDeploys: state.api.deploys.data,
+    shellCommandResponse: state.api.taskShellCommandResponse.data
+  };
+}
+
+export default connect(mapStateToProps)(TaskDetail);
```
Comment (andyhuang91) [inline SingularityUI/app/components/taskDetail/TaskDetail.jsx]: This file is pretty long. You might want to split it up into sub components.

Comment (tpetr) [inline SingularityUI/app/components/taskDetail/TaskDetail.jsx]: 👍

Comment (kwm4385) [inline SingularityUI/app/components/taskDetail/TaskDetail.jsx]: 🔪

**Comment by kwm4385** [comment]
Should be good to go @tpetr

**Pull request merged**